main.go

```diff
@@ -0,0 +1,257 @@
+		}
+
+		info, err = os.Lstat(pth)
+		if err != nil {
+			return err
+		}
+
+		header, err := zip.FileInfoHeader(info)
+		if err != nil {
+			return err
+		}
+
+		if baseDir != "" {
+			if isSymlink {
+				header.Name = filepath.Join(baseDir, strings.TrimPrefix(originalPath, sourcePath))
+			} else {
+				header.Name = filepath.Join(baseDir, strings.TrimPrefix(pth, sourcePath))
+			}
+
+		} else {
+			baseDir = "/"
+		}
+
+		if info.IsDir() {
+			header.Name += "/"
+		} else {
+			header.Method = zip.Deflate
+		}
+
+		writer, err := archive.CreateHeader(header)
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		file, err := os.Open(pth)
+		if err != nil {
+			return err
+		}
+
+		defer func() {
+			if err = file.Close(); err != nil {
+				log.Errorf("%s", err)
+			}
+		}()
+
+		_, err = io.Copy(writer, file)
+		return err
+	})
+	if err != nil {
+		return "", err
+	}
+
+	return fmt.Sprint(zipfile), nil
+}
+
+// If the file is a symbolic link it will evaulate the path name.
+
+// If the target is a symbolic link it will return true, the evaulated path, and the original path.
+// If the target is  not a symbolic link it will return false
+// If there is an error it will return the error.
+func checkSymlink(path string) (isSymlink bool, evaledPath string, error error) {
+	info, err := os.Lstat(path)
+	if err != nil {
+		return false, "", err
+	}
+
+	if info.Mode()&os.ModeSymlink == os.ModeSymlink {
+		evaledPath, err := filepath.EvalSymlinks(path)
+		if err != nil {
+			return true, "", err
+		}
+
+		return true, evaledPath, nil
+	}
+	return false, "", nil
+}
+
+// ensureDir will check the target path's existence.
+// If the targetDir does not exist it will create that.
+func ensureDir(targetDir string) {
+	dirOftargetPath := filepath.Dir(targetDir)
+
+	exits, err := pathutil.IsDirExists(dirOftargetPath)
+	if err != nil {
+		log.Errorf("Failed to open %s, error: %s", dirOftargetPath, err)
+	}
+
+	if exits == false {
+		fmt.Println()
+		log.Warnf("targetRootPath: %s does not exist", dirOftargetPath)
+		err = os.MkdirAll(dirOftargetPath, os.ModePerm)
+		if err != nil {
+			log.Errorf("Failed to create directory, error: %s", err)
+		}
+	}
+
+	checkAlreadyExist(targetDir)
+}
+
+func (configs config) print() {
+	log.Infof("Create ZIP configs:")
+	log.Printf("- SourcePath: %s", configs.SourcePath)
+	log.Printf("- TargetDir: %s", configs.TargetDir)
+}
+
+func (configs config) validate() error {
+	if err := input.ValidateIfNotEmpty(configs.SourcePath); err != nil {
+		return errors.New("issue with input SourcePath: " + err.Error())
+	}
+
+	if err := input.ValidateIfNotEmpty(configs.TargetDir); err != nil {
+		return errors.New("issue with input TargetDir: " + err.Error())
+	}
+
+	exist, err := pathutil.IsPathExists(configs.SourcePath)
+	if err != nil {
+		return err
+	}
+
+	if !exist {
+		return errors.New("issue with input SourcePath: ")
+	}
+
+	return nil
+}
+
+func fail(format string, v ...interface{}) {
```

Comment:
Rename it to `failf`, because the arguments are handled in the manner of `fmt.Printf`